File: src/main/java/com/openlattice/chronicle/serialization/ChronicleCallAdapterFactory.java
package com.openlattice.chronicle.serialization;

import com.google.common.base.Charsets;
import java.io.IOException;
import java.lang.annotation.Annotation;
import java.lang.reflect.Type;
import org.apache.commons.io.IOUtils;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import retrofit2.Call;
import retrofit2.CallAdapter;
import retrofit2.Response;
import retrofit2.Retrofit;

public class ChronicleCallAdapterFactory extends CallAdapter.Factory {
    private static final Logger logger = LoggerFactory.getLogger( ChronicleCallAdapterFactory.class );

    @Override
    public CallAdapter<?, ?> get( Type returnType, Annotation[] annotations, Retrofit retrofit ) {
        return new CallAdapter<Object, Object>() {
            @Override
            public Type responseType() {
                return returnType;
            }

            @Override public Object adapt( Call call ) {
                try {
                    Response response = call.execute();
                    if ( response.code() >= 400 ) {
                        logger.error( "Call failed with code {} and message {} and error body {}",
                                response.code(),
                                response.message(),
                                IOUtils.toString( response.errorBody().byteStream(), Charsets.UTF_8 ) );
                        return null;
                    }
                    return response.body();
                } catch ( IOException e ) {
                    logger.error( "Call failed.", e );
                    return null;
                }
            }

        };
    }

}

File: src/main/kotlin/com/openlattice/chronicle/serialization/ChronicleCallAdapterFactory.kt
package com.openlattice.chronicle.serialization

import com.google.common.base.Charsets
import org.apache.commons.io.IOUtils
import org.slf4j.LoggerFactory
import retrofit2.Call
import retrofit2.CallAdapter
import retrofit2.Retrofit
import java.io.IOException
import java.lang.reflect.Type

/**
 *
 * @author Matthew Tamayo-Rios &lt;matthew@getmethodic.com&gt;
 */
class ChronicleCallAdapterFactory : CallAdapter.Factory() {
    override fun get(returnType: Type, annotations: Array<Annotation>, retrofit: Retrofit): CallAdapter<*, *> {
        return object : CallAdapter<Any, Any> {
            override fun responseType(): Type {
                return returnType
            }

            override fun adapt(call: Call<Any>): Any? {
                return try {
                    val response = call.execute()
                    val code = response.code()
                    if (code >= 400) {
                        val body = IOUtils.toString(
                            response.errorBody()!!.byteStream(), Charsets.UTF_8
                        )
                        val url = call.request().url().toString()
                        val message = response.message()
                        val exMsg = "Call to $url failed with code $code and message $message"
                        logger.error(exMsg)
                        throw ChronicleCallException(exMsg, url, body, code)
                    }
                    response.body()
                } catch (e: IOException) {
                    logger.error("Call to ${call.request().url()} failed due to exception.", e)
                    throw e
                }
            }
        }
    }

    companion object {
        private val logger = LoggerFactory.getLogger(ChronicleCallAdapterFactory::class.java)
    }
}